### PR TITLE
Feature: Autocomplete Search Bar

### PR DESCRIPTION
**Pull request by PeytonBoggs**
This brach replaces the old Chakra <Input> search bar with a custom, open source <ReactSearchAutocomplete> component search bar. This component allows users to see up to 10 movies with titles similar to what they've input in the search bar. It also includes support for using the cursor or arrow keys to select a movie from the drop down menu. Once a user has typed in or selected the movie they want to guess, they can either press the guess button or hit enter to guess the movie. Currently, the autocomplete only shows the name of the movie, but this will be updated to include the year of release as well after that endpoint is merged in with the API.

https://github.com/Cineplexed/cineplexed-web/assets/129628668/1bf7ae30-ddbc-4c97-9c1d-16a54bd85e59

### CODE REVIEW

File: src/Game.tsx [hunk 1]
```diff
@@ -7,7 +7,7 @@ import { Movie } from "./Movie-Interface";
 import MysteryMovie from "./MysteryMovie";
 
 export default function Game() {
-    const [currentGuess, setCurrentGuess] = useState<Movie>({title: "Guessed Movie", year: "", gross: "", director: "", distributor: "", genres: [], actors: [], tagline: "", plot: "", poster: "",})
+    const [currentGuess, setCurrentGuess] = useState<Movie>({id: NaN, title: "Guessed Movie", year: "", gross: "", director: "", distributor: "", genres: [], actors: [], tagline: "", plot: "", poster: "",})
```
Comment (WesleyChenRTS2) [inline src/Game.tsx:10]: we can define an initial state variable outside of the component & use that here so it's less cluttered.

File: src/SearchBar.tsx [hunk 2]
```diff
@@ -10,31 +11,70 @@ interface SearchBarProps {
 }
 
 export default function SearchBar({ currentGuess, setCurrentGuess, guessList, setGuessList }: SearchBarProps) {
-    const [searchInput, setSearchInput] = useState<String>("")
+    const [searchOptions, setSearchOptions] = useState<MovieBite[]>([])
+    const [selectedMovie, setSelectedMovie] = useState<MovieBite>({id: NaN, title: "", year: ""})
```
Comment (WesleyChenRTS2) [inline src/SearchBar.tsx:15]: same here.

```
const initialSelectedMovieState = { ...initial state}

---

const [selectedMovie, setSelectedMovie] = useState<MovieBite>(initialSelectedMovieState)

```

File: src/SearchBar.tsx [hunk 3]
```diff
@@ -10,31 +11,70 @@ interface SearchBarProps {
...
-    
-    const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
-        setSearchInput(e.target.value)
-    }
 
-    function searchByInput() {
-        let optionsURL: string = "http://localhost:5050/getMovieOptions?title=" + searchInput
+    const handleInputChange = (input: string) => {
+        if (input === "") {
+            setSelectedMovie({id: NaN, title: "", year: ""})
```
Comment (WesleyChenRTS2) [inline src/SearchBar.tsx:20]:
```
setSelectedMovie(initialSelectedMovieState)
```

File: src/SearchBar.tsx [hunk 4]
```diff
@@ -10,31 +11,70 @@ interface SearchBarProps {
...
+            return
+        }
+        let optionsURL: string = "http://localhost:5050/getMovieOptions?title=" + input
```
Comment (WesleyChenRTS2) [inline src/SearchBar.tsx:23]: we don't want to hardcode urls like this, use environment variables to get localhost:5050, you can leave the rest.

File: src/SearchBar.tsx [hunk 5]
```diff
@@ -10,31 +11,70 @@ interface SearchBarProps {
...
         fetch(optionsURL)
         .then(response => response.json())
         .then(data => {
             if (data.results.length === 0) {
-                toast({
-                    title: "Movie not found",
-                    status: "error",
-                    position: "top-right"
-                })
+                setSelectedMovie({id: NaN, title: "", year: ""})
```
Comment (WesleyChenRTS2) [inline src/SearchBar.tsx:28]: same here

File: src/SearchBar.tsx [hunk 6]
```diff
@@ -47,38 +87,52 @@ export default function SearchBar({ currentGuess, setCurrentGuess, guessList, se
                     poster: "https://image.tmdb.org/t/p/w185" + data.poster_path
                 }
                 setCurrentGuess(formattedData)
-    
                 let tempGuessList: Movie[] = [...guessList]
                 tempGuessList.push(formattedData)
                 setGuessList(tempGuessList)
             })
-        })
     }
```
Comment (WesleyChenRTS2) [inline src/SearchBar.tsx:94]: GPT review that I agree with 

```// Extract baseURL for reuse and make the code cleaner.
const baseURL = 'http://localhost:5050';

// It's a good practice to have a separate function to handle API calls.
async function fetchApi(url, params) {
    const response = await fetch(`${baseURL}/${url}?${new URLSearchParams(params)}`);
    if (!response.ok) {
        throw new Error(`API request failed: ${response.statusText}`);
    }
    return await response.json();
}

// Extract details formatting into a separate function for better readability.
function formatMovieDetails(data) {
    return {
        title: data.title,
        year: data.release_date,
        gross: data.revenue,
        director: data.director,
        distributor: data.producer,
        genres: data.genres,
        actors: data.actors,
        tagline: data.tagline,
        plot: data.overview,
        poster: "https://image.tmdb.org/t/p/w185" + data.poster_path,
    };
}

async function searchByInput() {
    try {
        const options = await fetchApi('getMovieOptions', { title: searchInput });
        if (options.results.length === 0) {
            toast({
                title: "Movie not found",
                status: "error",
                position: "top-right",
            });
            return;
        }
        
        const details = await fetchApi('getMovieDetails', { id: options.results[0].id });
        const formattedData = formatMovieDetails(details);

        setCurrentGuess(formattedData);
        setGuessList(prevGuessList => [...prevGuessList, formattedData]);

    } catch (error) {
        console.error(error);
        // Handle any error that occurred during the API call
    }
}
```

**Review by WesleyChenRTS2 — APPROVED** [review]
with suggestions

File: src/SearchBar.tsx (hunk 4 above)
Comment (PeytonBoggs) [reply src/SearchBar.tsx:23]: Leaving this in for now - will change in next PR where all pointers will move from localhost to AWS. Noted about hardcoding URL, though

**Pull request merged**